### ISSUE TITLE
Potential bug in Utils::checkFilename preventing JSON uploads 

### ISSUE THREAD

**Issue by hamiltont** [getgrav/grav#2354]
When trying to upload a `.json` file to a page being written, I'm getting this error - `Unable to upload file barchart.json: Bad filename`. My (unmodified) `system/config/media.yaml` does contain an entry for json, and my `user/config/media.yaml` is empty.

I'm suspecting the issue is the below function from Utils (used in both the forms and admin plugin, and potentially others. The default grav install includes `.js` in the list of dangerous extensions, and I believe this is triggering a red herring, as a `.json` filename will include `.js`. I am having trouble confirming this - not sure how to log output from inside this function. 

```
    /**
     * Returns true if filename is considered safe.
     *
     * @param string $filename
     * @return bool
     */
    public static function checkFilename($filename)
    {
        $dangerous_extensions = Grav::instance()['config']->get('security.uploads_dangerous_extensions', []);
        array_walk($dangerous_extensions, function(&$val) {
            $val = '.' . $val;
        });

        $extension = '.' . pathinfo($filename, PATHINFO_EXTENSION);

        return !(
            // Empty filenames are not allowed.
            !$filename
            // Filename should not contain horizontal/vertical tabs, newlines, nils or back/forward slashes.
            || strtr($filename, "\t\v\n\r\0\\/", '_______') !== $filename
            // Filename should not start or end with dot or space.
            || trim($filename, '. ') !== $filename
            // Filename should not contain .php in it.
            || static::contains($extension, $dangerous_extensions)
        );
    }
``` 

For reference, here is the relevant portion of `system/config/security.yaml`. There are not local modifications to this file 

```
uploads_dangerous_extensions:
    - php
    - html
    - htm
    - js
    - exe
```

**Comment by rhukster** [comment]
Yup that 'contains' is catching `json` as a dangerous extension because of `js` in the list of dangerous extensions.

**Comment by mahagr** [comment]
There are two choices how to implement this properly. The easier way is to split the file by dot and check each individual section. The hard way is to build regexp from the listed extensions.

**Comment by hamiltont** [comment]
Side note - IMO the implementation may have been done this way (contains) because of stuff like `php5` - I'm not sure how to fix `.js != .json` without also losing that `.php == .php5`

**Comment by hamiltont** [comment]
A couple other potential edge cases: 

### Capitalization
as written, this won't check for capitalized filenames e.g. I can upload `foo.PHP`. It will fail, not because the filename check failed, but because "unknown file type", but if I want to upload `foo.PHP.jpg` would get past the filename check. Oddly, it will trigger a different error "The file is too large to be uploaded, maximum allowed is 8M according 
to your PHP settings. Increase your `post_max_size` PHP setting"

### No extension
I'm getting strange behavior trying to intentionally screw with the uploader. If I try to upload a file `foo"PHP` (no extension, because I know that internally PHP can often be tricked to turn a `"` into a `.`). This returns `Server responded with 0 code`

**Comment by wakqasahmed** [comment]
Also looks already fixed on `develop`. `Utils::checkFilename()` now extracts the extension via `pathinfo($filename, PATHINFO_EXTENSION)` and checks it against the dangerous-extensions list with `in_array()` (exact match), rather than the substring-style check that made `.json` collide with `.js`. A `.json` upload shouldn't trip the `.js` block anymore. Might be worth closing unless still reproducible.
